keygen/src/ssl.c

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright 2018 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <openssl/pem.h>
+#include <openssl/ec.h>
+#include <openssl/x509.h>
+#include <sodium.h>
+#include <errno.h>
+
+#include "ssl.h"
+#include "enftun/config.h"
+
+/**
+ * public_key_encode_x962() - Encode a public key from an EVP_PKEY structure
+ *
+ * Caller must free the returned buffer
+ *
+ * Return: A new buffer containing the x9.62-encoded string
+ */
+static char *public_key_encode_x962(EVP_PKEY *pkey)
+{
+    EC_KEY *key = NULL;
+    const EC_POINT * point = NULL;
+    char *hex = NULL;
+
+    key = EVP_PKEY_get1_EC_KEY(pkey);
+    if (!key) {
+        printf("%s Could not get EC_KEY from pkey", __func__);
+        goto out;
+    }
+
+    point = EC_KEY_get0_public_key(key);
+    if (!point) {
+        printf("%s Could not get point from EC_KEY", __func__);
+        goto out;
+    }
+
+    hex = EC_POINT_point2hex(EC_KEY_get0_group(key), point, EC_KEY_get_conv_form(key), NULL);
```

Comment:
Rather than use EC_KEY_get_conv_form, we should specify `POINT_CONVERSION_UNCOMPRESSED`.

I think the `xiam` documentation isn't clear on it, but I think somewhere (maybe in the XMB) we assume uncompressed x9.62 format.

Reply:
Changed.